File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int dx[4] = { 1, 0, -1, 0 }; 
int dy[4] = { 0, 1, 0, -1 };

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    
    for(vector<string> place: places) {        
        int flag = 1;
        for(int i = 0; i < 5; i++) {
            for(int j = 0; j < 5; j++) {
                if(place[i][j] != 'P') continue;
                
                queue<pair<int, pair<int, int>>> q; // {현재 이동 거리, {x, y}}
                bool visited[5][5] = { false };
                q.push({0, {i, j}});
                
                while(!q.empty()) {
                    auto curInfo = q.front();
                    q.pop();
                    
                    int dist = curInfo.first;
                    int curX = curInfo.second.first;
                    int curY = curInfo.second.second;
                    visited[curX][curY] = true;
                    
                    if(dist >= 2) continue; // 더 이상 어디 못감
                    
                    for(int dir = 0; dir < 4; dir++) {
                        int nxtX = curX + dx[dir];
                        int nxtY = curY + dy[dir];
                        
                        if(nxtX < 0 || nxtX >= 5 || nxtY < 0 || nxtY >= 5) continue;
                        if(place[nxtX][nxtY] == 'X') continue;
                        if(visited[nxtX][nxtY]) continue;
                        if(place[nxtX][nxtY] == 'P') {
                            flag = 0;
                            break;
                        }
                        
                        q.push({dist + 1, {nxtX, nxtY}});
                    }                    
                    if(!flag) break;
                }                
                if(!flag) break;
            }
            if(!flag) break;
        }
        
        answer.push_back(flag);
    }
    
    return answer;
}

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py
from itertools import product
from collections import deque

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]

def is_out(x, y):
    if x < 0 or x >= 5 or y < 0 or y >= 5:
        return True
    return False

def is_valid(board, x, y):
    q = deque()
    q.append((x, y, 0))
    visited = [[False] * 5 for _ in range(5)]
    visited[x][y] = True
    
    while q:
        curX, curY, curD = q.popleft()
        
        if curD >= 2:
            continue
        
        for dir in range(4):
            nxtX = curX + dx[dir]
            nxtY = curY + dy[dir]
            
            if is_out(nxtX, nxtY) or visited[nxtX][nxtY] or board[nxtX][nxtY] == 'X':
                continue
                
            if board[nxtX][nxtY] == 'P':
                break
            
            q.append((nxtX, nxtY, curD + 1))
        else:
            visited[curX][curY] = True
            continue
        break
    else:
        return True
    
    return False
        

def solution(places):
    answer = []
    
    for place in places:
        board = [[e for e in row] for row in place]
        for x, y in product(range(5), repeat=2):
            if board[x][y] == 'P' and not is_valid(board, x, y):
                break
        else:
            answer.append(1)
            continue
        answer.append(0)
                        
    return answer